0015-authority.md

```diff
@@ -0,0 +1,478 @@
+
+- An ACL system could potentially change how we think about ownership or authority or at least allow us to implement the authority system with much less code changes needed. Waiting for the ACL system could be a good idea.
+
+- Authority increases the spawn header size of a `NetworkObject` by 2 bit and the header size of `NetworkBehaviours` which use the `AllowAuthorityOverride` by 2 bit. In the future once we have more authority modes these headers will potentially get even larger.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+Authority could have been done as a permission/acl system instead of an authority mode. This would have allowed us to have more fine-grained control over who has authority. We were not able to find many real world use cases where this would be useful so that's why the authority mode was chosen. In the future we might consider storing authority internally as a set of owners instead of an authority mode but then most likely we would still want to expose authority as an authority mode for ease of use.
+
+# Prior art
+[prior-art]: #prior-art
+
+## UNet
+
+UNet also had the concept of [NetworkAuthority](https://docs.unity3d.com/Manual/UNetAuthority.html). There was a `Local Player Authority` checkbox on NetworkIdentity (MLAPI equivalent of NetworkObject). The UNet authority feature was very different from MLAPI's in UNet the authority was the one who was able to call commands (equivalent of server rpc). In UNet SyncVars (equivalent to NetworkVariable) where always controlled by the server, a client could never change them.
+
+The UNet authority system has a lot of problems. The NetworkTransform component was using authority because it used RPCs. Many user made components could not support authority because they wanted to have client driven SyncVars.
+
+This system is very limited and not really `authority` it is more what we have in MLAPI with `ownership`. It looks like the MLAPI `ownership` model has been inspired by the UNet `authority` model. The MLAPI `authority` model will improve MLAPI and allow users to do things which were not possible in UNet.
+
+## DOTS Netcode
+DOTS Netcode has a ghost mode which can be interpolated or predicted. The interpolated mode matches our 'server' mode. There is no mode for client authority. The predicted mode is something we do not support yet but can in the future.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Scrapping per NetworkBehaviour Authority
+See appendix for details. There are very little actual use cases for this and scrapping this will more then half the code size of the authority feature and lower the bandwidth usage of MLAPI.
```

Comment:
Is this something Gio could help with, finding whether there are use cases for this?

Reply:
@UnityGio for visibility